src/lib/query/pci.rs

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+
+use crate::{Iface, NisporError};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(into = "String")]
+#[serde(try_from = "String")]
+pub struct PciAddress {
+    pub domain: u32,
+    pub bus: u8,
+    pub slot: u8,
+    pub function: u8,
+}
+
+impl std::fmt::Display for PciAddress {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}:{:02x}:{:02x}.{}",
+            if self.domain > u16::MAX.into() {
+                format!("{:08x}", self.domain)
+            } else {
+                format!("{:04x}", self.domain)
+            },
+            self.bus,
+            self.slot,
+            self.function
+        )
+    }
+}
+
+impl TryFrom<String> for PciAddress {
+    type Error = NisporError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        Self::from_str(value.as_str())
+    }
+}
+
+impl From<PciAddress> for String {
+    fn from(v: PciAddress) -> String {
+        v.to_string()
+    }
+}
+
+impl FromStr for PciAddress {
+    type Err = NisporError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut items: Vec<&str> = s.split(":").collect();
+        let err = NisporError::invalid_argument(format!(
+            "Invalid PCI address, expecting string \
+                like 00:1b.0 or 0000:07:00.1, but got {s}"
+        ));
+        match items.len() {
+            3 | 2 => {
+                // It is safe to unwrap as we already checked the length.
+                let slot_function_str: Vec<&str> =
+                    items.pop().unwrap().split(".").collect();
+                if slot_function_str.len() != 2 {
+                    return Err(err.clone());
+                }
+
+                let slot = u8::from_str_radix(slot_function_str[0], 16)
+                    .map_err(|_| err.clone())?;
+
+                let function = slot_function_str[1]
+                    .parse::<u8>()
+                    .map_err(|_| err.clone())?;
+
+                let bus = u8::from_str_radix(items.pop().unwrap(), 16)
+                    .map_err(|_| err.clone())?;
+
+                let domain = if let Some(item) = items.pop() {
+                    u32::from_str_radix(item, 16).map_err(|_| err.clone())?
+                } else {
+                    0
+                };
+                Ok(Self {
+                    domain,
+                    bus,
+                    slot,
+                    function,
+                })
+            }
+            _ => Err(err),
+        }
+    }
+}
+
+impl Iface {
+    pub(crate) fn fill_pci_address(&mut self) {
+        if let Some(dev_path) =
+            PathBuf::from(format!("/sys/class/net/{}/device", self.name))
+                .read_link()
```

Comment:
Can you add a comment like https://github.com/nispor/nispor/blob/071a2c8c0e406b2248a3082961df7efa057d8aad/src/lib/query/iface.rs#L594?
It will be easier to find it and remember when the netlink interface becomes available.

Reply:
Added.